subsys/mgmt/fmfu/src/mgmt_fmfu.c

```diff
@@ -0,0 +1,281 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <logging/log.h>
+#include <mgmt/mgmt.h>
+#include <nrf_fmfu.h>
+#include <mgmt/mgmt_fmfu.h>
+#include "cborattr/cborattr.h"
+#include <stats/stats.h>
+
+LOG_MODULE_REGISTER(mgmt_mfu, CONFIG_MGMT_MFU_LOG_LEVEL);
+
+/* Define full modem update mgmt spesific SMP return codes */
+#define MGMT_ERR_EMODEM_INVALID_COMMAND 200
+#define MGMT_ERR_EMODEM_FAULT           201
+
+/* Define full modem mgmt commands */
+#define FMFU_MGMT_ID_GET_HASH 0
+#define FMFU_MGMT_ID_UPLOAD 1
+
+/* Define full modem IPC failures*/
+#define NRF_FMFU_RET_COMMAND_FAILED -3
+#define NRF_FMFU_RET_COMMAND_FAULT -4
+
+/* Struct used for parsing upload requests. */
+struct firmware_packet {
+	uint8_t *data_ptr;
+	uint8_t data[SMP_PACKET_MAXIMUM_TRANSMISSION_UNIT];
+	uint32_t target_address;
+	uint32_t data_len;
+};
+
+static int mgmt_get_memory_hash(struct mgmt_ctxt *ctxt);
+static int mgmt_firmware_upload(struct mgmt_ctxt *ctxt);
+static int unpack(struct mgmt_ctxt *ctxt, struct firmware_packet *packet, bool *whole_file_received);
+
+static const struct mgmt_handler mgmt_fmfu_handlers[] = {
+	[FMFU_MGMT_ID_GET_HASH] = {
+		.mh_read  = mgmt_get_memory_hash,
+		.mh_write = mgmt_get_memory_hash,
+	},
+	[FMFU_MGMT_ID_UPLOAD] = {
+		.mh_read  = NULL,
+		.mh_write = mgmt_firmware_upload
+	},
+};
+
+static struct mgmt_group fmfu_mgmt_group = {
+	.mg_handlers = (struct mgmt_handler *)mgmt_fmfu_handlers,
+	.mg_handlers_count = ARRAY_SIZE(mgmt_fmfu_handlers),
+	.mg_group_id = (MGMT_GROUP_ID_PERUSER + 1),
+};
+
+
+static int get_mgmt_err_from_modem_ret_err(int err)
+{
+	switch (err) {
+	case NRF_FMFU_RET_COMMAND_FAILED:
+		LOG_INF("RET_RPC_COMMAND_FAILED");
+		return MGMT_ERR_EMODEM_INVALID_COMMAND;
+	case NRF_FMFU_RET_COMMAND_FAULT:
+		LOG_INF("RET_RPC_COMMAND_FAULT");
+		return MGMT_ERR_EMODEM_FAULT;
+	default:
+		LOG_INF("BAD_STATE");
+		return MGMT_ERR_EBADSTATE;
+	}
+}
+
+int mgmt_fmfu_init(void)
+{
+	struct nrf_fmfu_digest digest;
+
+	LOG_INF("Initialized MFU mgmt module");
+	mgmt_register_group(&fmfu_mgmt_group);
+	int err = nrf_fmfu_init(&digest, NRF_FMFU_MODEM_BUFFER_SIZE,
+			(uint8_t *)0x20010000);
+	if (err != 0) {
+		LOG_ERR("nrf_fmfu_init failed: %d\n", err);
+		LOG_ERR("nrf_fmfu_init failed, errno: %d\n.", errno);
+		return err;
+	}
+	LOG_INF("Initialized MFU mgmt module done");
+	return err;
+}
+
+static int encode_response(struct mgmt_ctxt *ctx, uint32_t expected_offset)
+{
+	CborError err = 0;
+	int status = 0;
+
+	err |= cbor_encode_text_stringz(&ctx->encoder, "rc");
+	err |= cbor_encode_int(&ctx->encoder, status);
+	err |= cbor_encode_text_stringz(&ctx->encoder, "off");
+	err |= cbor_encode_uint(&ctx->encoder, expected_offset);
+
+	if (err != 0) {
+		return MGMT_ERR_ENOMEM;
+	}
+
+	return 0;
+}
+
+static int mgmt_firmware_upload(struct mgmt_ctxt *ctx)
+{
+
+	int ret;
+	int err;
+	bool whole_file_received;
+	static struct firmware_packet packet;
```

Comment:
Any particular reason why this is static?

Reply:
I guess it was just to save some stack in the function call as it doesn't need to be re-initialized every time. However I think it can safely be removed(premature optimization).